contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,300 @@
+
+This structure enables modularity. Many different types of tasks, including app-related ones, can run together. Work improving the code for the plugin server improves all of the processes. This also lowers the maintenance needed for having multiple systems.
+
+## Making sure arbitrary code doesn’t break everything
+
+The final piece to worry about is that apps are arbitrary code. When left unchecked, they can run whatever code they want, including code that tries to exploit or crash our servers. Allowing users to run arbitrary code can cause many security, infrastructure, and usability issues. We use VMs because they provide some solutions, but they don't offer total protection. We've done a bunch more work to prevent potential issues and make sure apps are secure and reliable.
+
+First, there are limitless libraries for JavaScript, but we only [allow a small portion of them](/docs/apps/build/reference#available-imports). You can’t install and use whatever npm package you like. We include basic packages such as Node’s standard `crypto`, `url`, `node-fetch`, and `zlib` libraries. We also include libraries like `snowflake-sdk`, `@google-cloud/bigquery`, and `pg` to connect external services. This ensures that the code run by our users and servers uses libraries we trust.
+
+Second, users can write a loop that loops forever and causes resource exhaustion. To prevent this, we set up a babel plugin that injects code into `for`, `while`, and `do while` loops (as described in [this article](https://medium.com/@bvjebin/js-infinite-loops-killing-em-e1c2f5f2db7f)). Whenever there is a loop, we set up a timer before we start it. If the loop runs for more than 30 seconds, the program errors.
+
+Third, to prevent data loss, we added the ability to retry logic to our `processEvent` function. Kafka handles retries for us by batching events and adding failed ones into the dead letter queue. We added the ability to run functions on `RetryError`  to improve consistency. App developers can also run fetches with retry. Here’s an example of what it looks like:
+
+```js
+import { RetryError } from '@posthog/plugin-scaffold'
+
+export function setupPlugin() {
+    try {
+        // Some network connection
+    } catch {
+        throw new RetryError('Service is unavailable, but it might be back up in a moment')
+    }
+}
+```
+
+Fourth, we made it easier to write high-quality apps. The biggest things we did here were creating an app [template](https://github.com/PostHog/posthog-plugin-starter-kit) and [scaffold](https://github.com/PostHog/plugin-scaffold) for users, including [Jest](https://jestjs.io/) to test the app, and the ability to write the app in TypeScript. Many more minor improvements added up to higher-quality apps getting written.
```

Comment:
This starter kit is quite out of date. It should at least be in TypeScript...

Also, what's "a scaffold"? How does one use it? The package you linked is some internal thing, which is used externally just for some TypeScript types

Reply:
Changed to docs, tutorials, TypeScript, and Jest. Scaffold now just linked as types, removed template.